smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/IdempotencyTokenIgnoredValidator.java

```diff
@@ -0,0 +1,131 @@
+import software.amazon.smithy.model.neighbor.Relationship;
+import software.amazon.smithy.model.neighbor.RelationshipType;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.traits.IdempotencyTokenTrait;
+import software.amazon.smithy.model.traits.MixinTrait;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+import software.amazon.smithy.utils.ListUtils;
+
+/**
+ * Emits warnings when a structure member has an idempotency token trait that will be ignored.
+ */
+public class IdempotencyTokenIgnoredValidator extends AbstractValidator {
+
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        if (!model.getAppliedTraits().contains(IdempotencyTokenTrait.ID)) {
+            return Collections.emptyList();
+        }
+        List<ValidationEvent> events = new ArrayList<>();
+        NeighborProvider reverse = NeighborProviderIndex.of(model).getReverseProvider();
+        for (MemberShape memberShape : model.getMemberShapes()) {
+            Shape container = model.expectShape(memberShape.getContainer());
+            // Skip non-structures (invalid) and mixins (handled at mixed site).
+            if (!container.isStructureShape() || container.hasTrait(MixinTrait.class)) {
+                continue;
+            }
+
+            if (memberShape.hasTrait(IdempotencyTokenTrait.class)) {
+                Trait trait = memberShape.expectTrait(IdempotencyTokenTrait.class);
+                events.addAll(checkRelationships(container.asStructureShape().get(), memberShape, trait, reverse));
+            }
+        }
+        return events;
+    }
+
+    private List<ValidationEvent> checkRelationships(
+        StructureShape containerShape,
+        MemberShape memberShape,
+        Trait trait,
+        NeighborProvider reverse
+    ) {
+
+        // Store relationships so we can emit one event per ignored binding.
+        Map<RelationshipType, List<ShapeId>> ignoredRelationships = new TreeMap<>();
+        List<ValidationEvent> events = new ArrayList<>();
+        List<Relationship> relationships = reverse.getNeighbors(containerShape);
+        int checkedRelationshipCount = relationships.size();
+        for (Relationship relationship : relationships) {
+            // Skip members of the container.
+            if (relationship.getRelationshipType() == RelationshipType.MEMBER_CONTAINER) {
+                checkedRelationshipCount--;
+                continue;
+            }
+            if (relationship.getRelationshipType() != RelationshipType.INPUT) {
+                ignoredRelationships.merge(relationship.getRelationshipType(),
```

Comment:
Can this be simplified to just use ignoredRelationships.computeIfAbsent? Might not need mergeShapeIdsList either at that point.